CMakeLists.txt

```diff
@@ -1,6 +1,33 @@
 cmake_minimum_required(VERSION 3.8)
 project(libcudacxx CXX)
 
+# Determine whether libcudacxx is the top-level project or included into
+# another project via add_subdirectory().
+if ("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_LIST_DIR}")
+  set(libcudacxx_TOPLEVEL_PROJECT ON)
+else()
+  set(libcudacxx_TOPLEVEL_PROJECT OFF)
+endif()
+
+include(cmake/libcudacxxInstallRules.cmake)
```

Comment:
We need to ensure that this file `lib/cmake/libcudacxx.cmake` is invoked each time libcudacxx is consumed via `add_subdirectory`

Consider when a project uses libcudacxx via CPM.

```
set(libcudacxx_ENABLE_INSTALL_RULES ON)
find_package(libcudacxx) #fails 
# cpm git clones libcudacxx
add_subdirectory(libcudacxx)
include(cmake/libcudacxxInstallRules.cmake) 
include(cmake/libcudacxxAddSubdir.cmake)
set(libcudacxx_DIR <path to libcudacxx-config.cmake> CACHE)
```

Now on first configuration this means we have install rules for libcudacxx. But if we reconfigure the project
this what happens:

```
set(libcudacxx_ENABLE_INSTALL_RULES ON)
find_package(libcudacxx) # works since libcudacxx_DIR is a cache var
```
Now we have no install rules when we do `cmake . && cmake . && cmake --install .` compared to we had install rules when we did `cmake . && cmake --install .`

Maybe after  `include(cmake/libcudacxxAddSubdir.cmake)` we unset the cache variable. We need to see if that works.

Reply:
We can just remove this when for non-toplevel usages to unblock. I only wanted this for testing Thrust installations, but we can find a better solution later when I start testing that. I don't think anyone else will want to add libcudacxx install rules to their builds.

Reply:
The last push moved the install rules to after the non-toplevel early return.

Reply:
When libcudacxx is consumed via `add_subdirectory` not having install rules means that the consumer could depend on a `libcudacxx::libcudacxx` target that doesn't exist. So when that project installs itself, it will have broken dependencies since `libcudacxx` hasn't been installed ( either the headers or the `libcudacxx-config.cmake` )

Reply:
I thought that this would be taken care of by the consumers export config. Let's discuss more when we chat later today.

Reply:
c7d9ddd implements the changes @robertmaynard and I discussed last week:

- Re-enable install rules for `add_subdirectory` builds
- Switch from `find_package` to explicitly including the `libcudacxx-config*` files, bypassing the `find_package` machinery that sets `libcudacxx_DIR` and breaks CPM.